lib/graphql-client.js

```diff
@@ -0,0 +1,63 @@
+const _ = require('lodash/fp')
+const https = require('https')
+const path = require('path')
+
+require('core-js/features/promise')
+const { ApolloClient, InMemoryCache, HttpLink, gql } = require('@apollo/client/core')
+const node_fetch = require('node-fetch')
+
+const fetchOptions = (() => {
+  const conf = (() => {
+    /* TODO: Sort of duplicated code from `brain.js`. */
+    const conf = require('./configuration').loadConfig({})
+    const dataPath = path.resolve(__dirname, '..', conf.brain.dataPath)
+    const resolve = file => path.resolve(dataPath, file)
+
+    return {
+      certPath: {
+        cert: resolve(conf.brain.certs.certFile),
+        key: resolve(conf.brain.certs.keyFile)
+      },
+      connectionInfoPath: resolve(conf.brain.connectionInfoPath)
+    }
+  })()
+
+  const pairing = require('./pairing')
+  const options = _.flow(
+    _.assign(pairing.getCert(conf.certPath)),
+    _.assign(pairing.connectionInfo(conf.connectionInfoPath))
+  )({ rejectUnauthorized: true })
+
+  return _.set('agent', https.Agent(options), options)
+})()
+
+const makeQueryString = _.flow(
+  _.toPairs,
+  _.map(_.join('=')),
+  _.join('&')
+)
+
+const fetch = (uri, options) =>
+  node_fetch(uri, _.update('headers', _.set('date', new Date().toISOString()), options))
+
+/*
+ * opts is a { key: val, ... } object that will be turned into the
+ * `key=val&...` query string
+ */
+const GraphQLClient = opts =>
+  new ApolloClient({
+    cache: new InMemoryCache(),
+    defaultOptions: {
+      query: {
+        fetchPolicy: 'no-cache',
+        errorPolicy: 'all',
+      }
+    },
+    link: new HttpLink({
+      uri: 'https://localhost:3000/graphql?' + makeQueryString(opts),
```

Comment:
Shouldn't the `?` only appear if `opts` is not empty or null?

Reply:
I think that the query string of an URL can be empty (would have to check to make sure). But the client is only ever used in the trader right now so it won't ever be empty/null. If the client is to be used in the future on more places then this should be changed, yes.